File: scripts/fetch-data.py
"""
옵티멈스터디카페 네이버 플레이스 데이터 수집 스크립트
- Place ID: 1333416518
- GraphQL API + REST API로 매장 정보, 사진 수집
- 결과를 data/place-data.json에 저장
"""
import requests
import json
import os
import time
import random
import urllib.request

PLACE_ID = "1333416518"
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(BASE_DIR, "data")
GALLERY_DIR = os.path.join(BASE_DIR, "public", "images", "gallery")

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Referer": "https://m.place.naver.com/",
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "ko-KR,ko;q=0.9",
}

PLACE_API_URL = "https://pcmap-api.place.naver.com/graphql"


def fetch_via_graphql():
    """GraphQL API로 매장 상세 정보 수집"""
    payload = [
        {
            "operationName": "getPlaceDetail",
            "variables": {"id": PLACE_ID},
            "query": """
            query getPlaceDetail($id: String!) {
                placeDetail(id: $id) {
                    id
                    name
                    roadAddress
                    address
                    phone
                    category
                    visitorReviewCount
                    blogReviewCount
                    visitorReviewScore
                    businessHours
                    menuInfo {
                        menuList {
                            name
                            price
                        }
                    }
                    facilityInfo {
                        facilityList
                    }
                    imageUrl
                    description
                    x
                    y
                }
            }
            """
        }
    ]

    try:
        resp = requests.post(
            PLACE_API_URL,
            json=payload,
            headers={**HEADERS, "Content-Type": "application/json"},
            timeout=15,
        )
        if resp.status_code == 200:
            data = resp.json()
            if isinstance(data, list) and data:
                place = data[0].get("data", {}).get("placeDetail")
                if place:
                    return parse_place_detail(place)
        print(f"[GraphQL] 응답 코드: {resp.status_code}")
    except Exception as e:
        print(f"[GraphQL 오류] {e}")
    return None


def fetch_via_rest_api():
    """REST API로 매장 정보 수집 (fallback)"""
    result = {}

    # 기본 정보
    try:
        url = f"https://map.naver.com/p/api/sites/summary/{PLACE_ID}?lang=ko"
        resp = requests.get(url, headers=HEADERS, timeout=15)
        if resp.status_code == 200:
            data = resp.json()
            result.update({
                "name": data.get("name", ""),
                "address": data.get("roadAddress", "") or data.get("address", ""),
                "telephone": data.get("phone", ""),
                "category": data.get("category", ""),
                "rating": data.get("averageRating") or data.get("visitorReviewScore") or 0,
                "visitor_review_count": data.get("visitorReviewCount", 0),
                "blog_review_count": data.get("blogCafeReviewCount", 0),
                "business_hours": data.get("businessHours", ""),
                "description": data.get("description", ""),
                "x": data.get("x"),
                "y": data.get("y"),
            })
    except Exception as e:
        print(f"[REST 기본정보 오류] {e}")

    time.sleep(random.uniform(0.5, 1.0))

    # 메뉴/요금 정보
    try:
        url = f"https://map.naver.com/p/api/sites/summary/{PLACE_ID}/menu?lang=ko"
        resp = requests.get(url, headers=HEADERS, timeout=15)
        if resp.status_code == 200:
            data = resp.json()
            menus = []
            if isinstance(data, list):
                menus = data
            elif isinstance(data, dict) and "menus" in data:
                menus = data["menus"]
            result["price_info"] = [
                {"name": m.get("name", ""), "price": m.get("price", "")}
                for m in menus[:30] if m.get("name")
            ]
    except Exception as e:
        print(f"[REST 메뉴 오류] {e}")

    time.sleep(random.uniform(0.5, 1.0))

    # 시설 정보
    try:
        url = f"https://map.naver.com/p/api/sites/summary/{PLACE_ID}/detailed?lang=ko"
        resp = requests.get(url, headers=HEADERS, timeout=15)
        if resp.status_code == 200:
            data = resp.json()
            facilities = data.get("facilities", [])
            if data.get("convenience"):
                facilities += data["convenience"]
            result["facilities"] = facilities
    except Exception as e:
        print(f"[REST 시설 오류] {e}")

    return result if result.get("name") else None


def fetch_photos():
    """네이버 플레이스 사진 URL 수집"""
    photos = []
    try:
        url = f"https://map.naver.com/p/api/sites/summary/{PLACE_ID}/photo?lang=ko"
        resp = requests.get(url, headers=HEADERS, timeout=15)
        if resp.status_code == 200:
            data = resp.json()
            if isinstance(data, dict):
                photo_list = data.get("photos", data.get("list", []))
            elif isinstance(data, list):
                photo_list = data
            else:
                photo_list = []

            for item in photo_list[:20]:
                if isinstance(item, str):
                    photos.append(item)
                elif isinstance(item, dict):
                    photo_url = item.get("url") or item.get("imageUrl") or item.get("thumbnailUrl") or ""
                    if photo_url:
                        photos.append(photo_url)
        print(f"[사진] {len(photos)}장 수집")
    except Exception as e:
        print(f"[사진 수집 오류] {e}")
    return photos


def download_images(photo_urls):
    """사진을 로컬에 다운로드"""
    downloaded = []
    os.makedirs(GALLERY_DIR, exist_ok=True)

    for i, url in enumerate(photo_urls):
        try:
            filename = f"gallery_{i+1:02d}.jpg"
            filepath = os.path.join(GALLERY_DIR, filename)
            req = urllib.request.Request(url, headers={"User-Agent": HEADERS["User-Agent"]})
            with urllib.request.urlopen(req, timeout=15) as response:
                with open(filepath, "wb") as f:
                    f.write(response.read())
            downloaded.append(f"/images/gallery/{filename}")
            print(f"  다운로드: {filename}")
        except Exception as e:
            print(f"  다운로드 실패 ({url[:50]}...): {e}")
            # CDN URL을 fallback으로 유지
            downloaded.append(url)

    return downloaded


def parse_place_detail(data):
    """API 응답을 표준 형식으로 변환"""
    price_info = []
    menu_info = data.get("menuInfo") or {}
    for menu in menu_info.get("menuList", []):
        price_info.append({
            "name": menu.get("name", ""),
            "price": menu.get("price", ""),
        })

    facilities = []
    facility_info = data.get("facilityInfo") or {}
    facilities = facility_info.get("facilityList", [])

    return {
        "name": data.get("name", ""),
        "address": data.get("roadAddress", "") or data.get("address", ""),
        "telephone": data.get("phone", ""),
        "category": data.get("category", ""),
        "visitor_review_count": data.get("visitorReviewCount", 0) or 0,
        "blog_review_count": data.get("blogReviewCount", 0) or 0,
        "rating": data.get("visitorReviewScore", 0) or 0,
        "business_hours": data.get("businessHours", ""),
        "price_info": price_info,
        "facilities": facilities if isinstance(facilities, list) else [],
        "description": data.get("description", ""),
        "x": data.get("x"),
        "y": data.get("y"),
    }


def main():
    print("=" * 50)
    print("옵티멈스터디카페 데이터 수집 시작")
    print(f"Place ID: {PLACE_ID}")
    print("=" * 50)

    # 1. 매장 정보 수집
    print("\n[1] GraphQL API로 매장 정보 수집 중...")
    detail = fetch_via_graphql()

    if not detail:
        print("[1] GraphQL 실패, REST API fallback...")
        time.sleep(random.uniform(1.0, 2.0))
        detail = fetch_via_rest_api()

    if not detail:
        print("[오류] 매장 정보를 수집하지 못했습니다.")
        # 기본 정보로 fallback
        detail = {
            "name": "옵티멈스터디카페",
            "address": "서울 관악구 신림로15길 31",
            "telephone": "0507-1365-5966",
            "category": "스터디카페",
            "visitor_review_count": 0,
            "blog_review_count": 0,
            "rating": 0,
            "business_hours": "24시간 연중무휴",
            "price_info": [],
            "facilities": [],
            "description": "신림동 최초 단독 건물형 스터디카페. 300평 5개층 220석.",
            "x": "126.9285",
            "y": "37.4687",
        }
        print("[기본 정보로 대체합니다]")

    print(f"  매장명: {detail.get('name')}")
    print(f"  주소: {detail.get('address')}")
    print(f"  요금 항목 수: {len(detail.get('price_info', []))}")
    print(f"  편의시설 수: {len(detail.get('facilities', []))}")

    # 2. 사진 수집
    print("\n[2] 사진 수집 중...")
    time.sleep(random.uniform(1.0, 2.0))
    photo_urls = fetch_photos()

    # 3. 이미지 다운로드
    local_images = []
    if photo_urls:
        print(f"\n[3] {len(photo_urls)}장 이미지 다운로드 중...")
        local_images = download_images(photo_urls)
    else:
        print("\n[3] 수집된 사진이 없습니다.")

    # 4. JSON 저장
    result = {
        **detail,
        "place_id": PLACE_ID,
        "images": local_images,
        "image_urls_original": photo_urls,
        "naver_place_url": f"https://map.naver.com/p/entry/place/{PLACE_ID}",
        "kakao_channel": "https://pf.kakao.com/_vxgxjIn",
        "blog_url": "https://blog.naver.com/optimumstudycafe",
        "instagram": "https://www.instagram.com/opt_studycafe/",
    }

    os.makedirs(DATA_DIR, exist_ok=True)
    output_path = os.path.join(DATA_DIR, "place-data.json")
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(result, f, ensure_ascii=False, indent=2)

    print(f"\n[완료] 데이터 저장: {output_path}")
    print(f"  이미지: {len(local_images)}장")
    print("=" * 50)


if __name__ == "__main__":
    main()

File: scripts/fetch-data.mjs
/**
 * 옵티멈스터디카페 네이버 플레이스 데이터 수집 스크립트 (Node.js)
 * - Place ID: 1333416518
 * - REST API로 매장 정보, 사진 수집
 * - 결과를 data/place-data.json에 저장
 */
import https from "https";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BASE_DIR = path.resolve(__dirname, "..");
const DATA_DIR = path.join(BASE_DIR, "data");
const GALLERY_DIR = path.join(BASE_DIR, "public", "images", "gallery");
const PLACE_ID = "1333416518";

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
  Referer: "https://m.place.naver.com/",
  Accept: "application/json, text/plain, */*",
  "Accept-Language": "ko-KR,ko;q=0.9",
};

function httpsGet(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: "GET",
      headers: { ...HEADERS, ...headers },
    };
    const req = https.request(options, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        httpsGet(res.headers.location, headers).then(resolve).catch(reject);
        return;
      }
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve(data));
    });
    req.on("error", reject);
    req.end();
  });
}

function httpsPost(url, body, headers = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const postData = JSON.stringify(body);
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: "POST",
      headers: {
        ...HEADERS,
        ...headers,
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(postData),
      },
    };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve(data));
    });
    req.on("error", reject);
    req.write(postData);
    req.end();
  });
}

function downloadFile(url, filepath) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      headers: { "User-Agent": HEADERS["User-Agent"] },
    };
    https
      .get(options, (res) => {
        if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          downloadFile(res.headers.location, filepath).then(resolve).catch(reject);
          return;
        }
        const ws = fs.createWriteStream(filepath);
        res.pipe(ws);
        ws.on("finish", () => {
          ws.close();
          resolve(true);
        });
      })
      .on("error", reject);
  });
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function fetchViaGraphQL() {
  console.log("  GraphQL API 호출 중...");
  const payload = [
    {
      operationName: "getPlaceDetail",
      variables: { id: PLACE_ID },
      query: `
        query getPlaceDetail($id: String!) {
          placeDetail(id: $id) {
            id name roadAddress address phone category
            visitorReviewCount blogReviewCount visitorReviewScore
            businessHours
            menuInfo { menuList { name price } }
            facilityInfo { facilityList }
            imageUrl description x y
          }
        }
      `,
    },
  ];

  try {
    const raw = await httpsPost("https://pcmap-api.place.naver.com/graphql", payload);
    const data = JSON.parse(raw);
    if (Array.isArray(data) && data[0]?.data?.placeDetail) {
      const p = data[0].data.placeDetail;
      return parsePlaceDetail(p);
    }
  } catch (e) {
    console.log(`  GraphQL 오류: ${e.message}`);
  }
  return null;
}

async function fetchViaRestAPI() {
  console.log("  REST API fallback...");
  const result = {};

  // 기본 정보
  try {
    const raw = await httpsGet(`https://map.naver.com/p/api/sites/summary/${PLACE_ID}?lang=ko`);
    const data = JSON.parse(raw);
    Object.assign(result, {
      name: data.name || "",
      address: data.roadAddress || data.address || "",
      telephone: data.phone || "",
      category: data.category || "",
      rating: data.averageRating || data.visitorReviewScore || 0,
      visitor_review_count: data.visitorReviewCount || 0,
      blog_review_count: data.blogCafeReviewCount || 0,
      business_hours: data.businessHours || "",
      description: data.description || "",
      x: data.x,
      y: data.y,
    });
  } catch (e) {
    console.log(`  기본 정보 오류: ${e.message}`);
  }

  await sleep(800);

  // 메뉴/요금
  try {
    const raw = await httpsGet(`https://map.naver.com/p/api/sites/summary/${PLACE_ID}/menu?lang=ko`);
    const data = JSON.parse(raw);
    const menus = Array.isArray(data) ? data : data.menus || [];
    result.price_info = menus
      .slice(0, 30)
      .map((m) => ({ name: m.name || m.menuName || "", price: m.price || m.unitPrice || "" }))
      .filter((m) => m.name);
  } catch (e) {
    console.log(`  메뉴 정보 오류: ${e.message}`);
  }

  await sleep(800);

  // 시설 정보
  try {
    const raw = await httpsGet(`https://map.naver.com/p/api/sites/summary/${PLACE_ID}/detailed?lang=ko`);
    const data = JSON.parse(raw);
    result.facilities = [...(data.facilities || []), ...(data.convenience || [])];
  } catch (e) {
    console.log(`  시설 정보 오류: ${e.message}`);
  }

  return result.name ? result : null;
}

async function fetchPhotos() {
  const photos = [];
  try {
    const raw = await httpsGet(`https://map.naver.com/p/api/sites/summary/${PLACE_ID}/photo?lang=ko`);
    const data = JSON.parse(raw);
    const list = Array.isArray(data) ? data : data.photos || data.list || [];
    for (const item of list.slice(0, 20)) {
      const url = typeof item === "string" ? item : item.url || item.imageUrl || item.thumbnailUrl || "";
      if (url) photos.push(url);
    }
    console.log(`  사진 ${photos.length}장 수집`);
  } catch (e) {
    console.log(`  사진 수집 오류: ${e.message}`);
  }
  return photos;
}

async function downloadImages(photoUrls) {
  fs.mkdirSync(GALLERY_DIR, { recursive: true });
  const downloaded = [];

  for (let i = 0; i < photoUrls.length; i++) {
    const filename = `gallery_${String(i + 1).padStart(2, "0")}.jpg`;
    const filepath = path.join(GALLERY_DIR, filename);
    try {
      await downloadFile(photoUrls[i], filepath);
      downloaded.push(`/images/gallery/${filename}`);
      console.log(`  다운로드: ${filename}`);
    } catch (e) {
      console.log(`  다운로드 실패: ${e.message}`);
      downloaded.push(photoUrls[i]);
    }
  }
  return downloaded;
}

function parsePlaceDetail(data) {
  const priceInfo = [];
  const menuList = data.menuInfo?.menuList || [];
  for (const menu of menuList) {
    priceInfo.push({ name: menu.name || "", price: menu.price || "" });
  }

  const facilities = data.facilityInfo?.facilityList || [];

  return {
    name: data.name || "",
    address: data.roadAddress || data.address || "",
    telephone: data.phone || "",
    category: data.category || "",
    visitor_review_count: data.visitorReviewCount || 0,
    blog_review_count: data.blogReviewCount || 0,
    rating: data.visitorReviewScore || 0,
    business_hours: data.businessHours || "",
    price_info: priceInfo,
    facilities: Array.isArray(facilities) ? facilities : [],
    description: data.description || "",
    x: data.x,
    y: data.y,
  };
}

async function main() {
  console.log("==================================================");
  console.log("옵티멈스터디카페 데이터 수집 시작");
  console.log(`Place ID: ${PLACE_ID}`);
  console.log("==================================================");

  // 1. 매장 정보
  console.log("\n[1] 매장 정보 수집 중...");
  let detail = await fetchViaGraphQL();
  if (!detail) {
    await sleep(1500);
    detail = await fetchViaRestAPI();
  }

  if (!detail) {
    console.log("[기본 정보로 대체합니다]");
    detail = {
      name: "옵티멈스터디카페",
      address: "서울 관악구 신림로15길 31",
      telephone: "0507-1365-5966",
      category: "스터디카페",
      visitor_review_count: 0,
      blog_review_count: 0,
      rating: 0,
      business_hours: "24시간 연중무휴",
      price_info: [],
      facilities: [],
      description: "신림동 최초 단독 건물형 스터디카페. 300평 5개층 220석.",
      x: "126.9285",
      y: "37.4687",
    };
  }

  console.log(`  매장명: ${detail.name}`);
  console.log(`  주소: ${detail.address}`);
  console.log(`  요금 항목: ${detail.price_info?.length || 0}개`);
  console.log(`  편의시설: ${detail.facilities?.length || 0}개`);

  // 2. 사진 수집
  console.log("\n[2] 사진 수집 중...");
  await sleep(1500);
  const photoUrls = await fetchPhotos();

  // 3. 이미지 다운로드
  let localImages = [];
  if (photoUrls.length > 0) {
    console.log(`\n[3] ${photoUrls.length}장 이미지 다운로드 중...`);
    localImages = await downloadImages(photoUrls);
  } else {
    console.log("\n[3] 수집된 사진이 없습니다.");
  }

  // 4. JSON 저장
  const result = {
    ...detail,
    place_id: PLACE_ID,
    images: localImages,
    image_urls_original: photoUrls,
    naver_place_url: `https://map.naver.com/p/entry/place/${PLACE_ID}`,
    kakao_channel: "https://pf.kakao.com/_vxgxjIn",
    blog_url: "https://blog.naver.com/optimumstudycafe",
    instagram: "https://www.instagram.com/opt_studycafe/",
    fetchedAt: new Date().toISOString(),
  };

  fs.mkdirSync(DATA_DIR, { recursive: true });
  const outputPath = path.join(DATA_DIR, "place-data.json");
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2), "utf-8");

  console.log(`\n[완료] 데이터 저장: ${outputPath}`);
  console.log(`  이미지: ${localImages.length}장`);
  console.log("==================================================");
}

main().catch(console.error);
